[Merged by Bors] - feat: add Dockerfiles for lean and mathlib

zeramorphic opened the pull request:
Add Dockerfiles for lean and mathlib

---

Based off https://github.com/leanprover-community/mathlib3/tree/master/.docker. Also modifies the current gitpod Dockerfile so that leanblueprint works out-of-the-box.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies reviewed (approved):
I have just updated [our Dockerhub image](https://hub.docker.com/repository/docker/leanprovercommunity/) using the two new scripts in this PR. Everything works as intended when I open the LeanAPAP repo (which points to the `leanprovercommunity/mathlib:gitpod` image).

@eric-wieser, can you make sure you too are happy with the changes?

eric-wieser commented on .docker/gitpod/Dockerfile:
```diff
@@ -1,12 +1,16 @@
 
-RUN apt-get update && apt-get install sudo git curl git bash-completion python3 python3-requests -y && apt-get clean
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get install sudo git curl git bash-completion texlive texlive-xetex latexmk graphviz graphviz-dev python3 python3-pip python3-requests -y && apt-get clean
```
Why do we want `texlive` for mathlib? This makes the image quite a bit bigger.

zeramorphic replied:
Maybe we should make two tags: one that includes a copy of leanblueprint and its dependencies (such as TeXLive), and one that does not.

YaelDillies replied:
This is because we want it for blueprints. But maybe that should be an extra tag/dockerhub repo

zeramorphic commented:
I added a -4 suffix to the names of the images, and added automatic date tags.
![image](https://github.com/user-attachments/assets/f0643157-5c0e-433b-90f6-392f5b4c5572)

YaelDillies commented:
I think we should get this in now. Our dockerhub is already using these new images

maintainer merge

bryangingechen commented on scripts/docker_push.sh:
```diff
@@ -0,0 +1,15 @@
+# This script attempts to build all the docker images, and then push then to the repository.
+# You'll need to have run `docker login` already;
+# check with scott.morrison@gmail.com for credentials.
```
Maybe
```suggestion
# check with kim@tqft.net for credentials.
```
@kim-em Do you want your email here at all? Maybe just pointing to Zulip is enough?

kim-em replied:
```suggestion
# check with Mathlib maintainers on Zulip for credentials.
```

bryangingechen commented:
Also, please merge `master` and document the two new scripts in `scripts/README.md`, per our new linter in #14539.

kim-em commented:
bors merge

kim-em commented:
bors merge
